### PR TITLE
Track external protocol work in LangevinSplittingIntegrator

### PR DESCRIPTION
**Pull request by bas-rustenburg**
Why:

Some codebases, for instance protons and saltswap, modify context
parameters externally. To save expensive energy computations,
we want to accumulate the protocol work inside the integrator.
This commit adds a `protocol_work` variable to keep track of it.

It also adds a unit test to make sure the new feature works adequately.

This change addresses the need by:

* Adding the `measure_protocol_work` parameter, default false, and
including the machinery that calculates it.
* Adding a unit test to ensure it works as expected.

Please double-check that the tests passes before merging. I haven't been able to run it locally yet.

### CODE REVIEW

**Comment by pgrinaway** [comment]
🔥  nice

**Comment by bas-rustenburg** [comment]
It would be appreciated if @maxentile and @gregoryross take a look at this to make sure the implementation is correct, since they should be most familiar with the integrator and procedure respectively.

**Comment by bas-rustenburg** [comment]
Thanks Josh, I'll fix that. It uses the energy at the end of the _previous_ step, and at the start of the upcoming step.

File: openmmtools/tests/test_integrators.py [hunk 1]
```diff
@@ -206,6 +206,39 @@ def test_vvvr_shadow_work_accumulation():
...
+    context = openmm.Context(system, integrator)
+    context.setPositions(testsystem.positions)
+    context.setVelocitiesToTemperature(temperature)
+    assert(integrator.getGlobalVariableByName('protocol_work') == 0), "Protocol work should be 0 initially"
+    integrator.step(25)
+    assert(integrator.getGlobalVariableByName('protocol_work') == 0), "There should be no protocol work."
+    perturbed_K=99.0 * unit.kilocalories_per_mole / unit.angstroms**2
+    context.setParameter('testsystems_HarmonicOscillator_K', perturbed_K)
+    integrator.step(1)
+    assert (integrator.getGlobalVariableByName('protocol_work') != 0), "There should be protocol work after perturbing."
```
Comment (gregoryross) [inline openmmtools/tests/test_integrators.py]: There are multiple `assert` statements in this functions that each test a different aspect. It's certainly worth considering splitting these into separate tests.

Comment (bas-rustenburg) [inline openmmtools/tests/test_integrators.py]: Perhaps not optimal. I was just following the original test structure that Josh used. It should at least fail with a specific error message if it does fail.

File: openmmtools/tests/test_integrators.py [hunk 2]
```diff
@@ -206,6 +206,39 @@ def test_vvvr_shadow_work_accumulation():
     assert('shadow_work' not in names_of_globals)
 
 
+def test_vvvr_protocol_work_accumulation():
```
Comment (gregoryross) [inline openmmtools/tests/test_integrators.py]: I strongly recommend testing the accumulated work explicitly by comparing with `getState(getEnergy=True)`. I *thought* I had a test for this in `saltswap` here:

https://github.com/choderalab/saltswap/blob/master/saltswap/tests/test_integrators.py#L185-L206

But it's not exactly the same as testing the internally accumulated work. Anyway, something that tests for that would be good. (I'll certainly be adding to my `saltswap` tests.)

Comment (bas-rustenburg) [inline openmmtools/tests/test_integrators.py]: I'll add that, thanks for the suggestion!

**Comment by bas-rustenburg** [comment]
Just put in a new commit that hopefully addresses @maxentile's comments.

File: openmmtools/integrators.py [hunk 3]
```diff
@@ -1093,14 +1098,31 @@ def substep_function(step_string):
...
         # Integrate
         self.addUpdateContextState()
         self.addComputeTemperatureDependentConstants({"sigma": "sqrt(kT/m)"})
+
+        # Protocol work is calculated by taking the potential energy in the perturbed system
+        # and subtracting the previous, unperturbed potential energy
+        # before any steps are taken.
+        if measure_protocol_work:
+            self.addComputeGlobal("perturbed_pe", "energy")
+            self.addComputeGlobal("protocol_work", "protocol_work + (perturbed_pe - new_pe)")
```
Comment (gregoryross) [inline openmmtools/integrators.py]: I find it difficult to wrap my head around `CustomIntegrators` at the best of times, but is there potential for conflict if `measure_shadow_work = True`, as `new_pe` is set to zero?

Comment (bas-rustenburg) [inline openmmtools/integrators.py]: You might be right there. Latest fix takes care of it.

**Comment by bas-rustenburg** [comment]
Fixed Greg's concern and no longer using new_pe, thanks for suggesting it Josh!

**Comment by maxentile** [comment]
Looks great to me!

**Comment by bas-rustenburg** [comment]
Update to test against external potential energy included now.

**Comment by bas-rustenburg** [comment]
Bugfix

**Comment by bas-rustenburg** [comment]
This should hopefully be the last bug. It passes for me locally now.

**Comment by bas-rustenburg** [comment]
I hereby declare this PR, merged!

**Pull request merged**